src/execution_plan/execution_plan.c

```diff
@@ -337,25 +346,25 @@ static void _ExecutionPlanSegment_ProcessQueryGraph(ExecutionPlanSegment *segmen
 
 		if(connectedComponentsCount > 1) {
 			// Connect traversal operations.
-			OpBase *childOp;
-			OpBase *parentOp;
-			Vector_Pop(traversals, &parentOp);
+			OpBase *parentOp = array_pop(traversals);
 			// Connect cartesian product to the root of traversal.
 			ExecutionPlan_AddOp(cartesianProduct, parentOp);
-			while(Vector_Pop(traversals, &childOp)) {
+			uint count = array_len(traversals);
+			for(uint j = 0; j < count; j ++) {
+				OpBase *childOp = array_pop(traversals);
 				ExecutionPlan_AddOp(parentOp, childOp);
 				parentOp = childOp;
 			}
 		} else {
-			for(int traversalIdx = 0; traversalIdx < Vector_Size(traversals); traversalIdx++) {
-				Vector_Get(traversals, traversalIdx, &op);
-				Vector_Push(ops, op);
+			uint count = array_len(traversals);
+			for(uint j = 0; j < count; j++) {
+				_ExecutionPlan_UpdateRoot(segment, traversals[j]);
```

Comment:
This is OK, another option would be to chain traversals `0 .. j` and call `_ExecutionPlan_UpdateRoot` once with traversal `j`